src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp

```diff
@@ -82,315 +99,510 @@ enum FourHorsemen
+            if (instance->GetBossState(BOSS_HORSEMEN) == IN_PROGRESS)
+                return;
+            if (!instance->CheckRequiredBosses(BOSS_HORSEMEN))
+                return ResetEncounter();
+            instance->SetBossState(BOSS_HORSEMEN, IN_PROGRESS);
+            Map::PlayerList const &players = me->GetMap()->GetPlayers();
+            if (players.isEmpty()) // sanity check
+                ResetEncounter();
+
+            for (Horseman boss : horsemen)
+                if (Creature* cBoss = getHorsemanHandle(boss))
+                {
+                    if (!cBoss->IsAlive())
+                        return ResetEncounter();
+                    cBoss->SetReactState(REACT_PASSIVE);
+                    cBoss->AttackStop(); // clear initial target that was set on enter combat
+                    cBoss->setActive(true);
+
+                    for (Map::PlayerList::const_iterator it = players.begin(); it != players.end(); ++it)
+                        if (Player* player = it->GetSource())
+                        {
+                            if (player->IsGameMaster())
+                                continue;
+
+                            if (player->IsAlive())
+                            {
+                                cBoss->AddThreat(player, 0.0f);
+                                cBoss->SetInCombatWith(player);
+                                player->SetInCombatWith(cBoss);
+                            }
+                        }
 
-            if (Thane && Lady && Baron && Sir)
-            {
-                if (attack && who)
+                    /* Why do the Four Horsemen run to opposite corners of the room when engaged?          *
+                    * They saw all the mobs leading up to them being AoE'd down and made a judgment call. */
+                    cBoss->AI()->DoAction(ACTION_BEGIN_MOVEMENT);
+                }
+                else
                 {
-                    ENSURE_AI(boss_four_horsemen::boss_four_horsemenAI, Thane->AI())->encounterActionAttack = true;
-                    ENSURE_AI(boss_four_horsemen::boss_four_horsemenAI, Lady->AI())->encounterActionAttack = true;
-                    ENSURE_AI(boss_four_horsemen::boss_four_horsemenAI, Baron->AI())->encounterActionAttack = true;
-                    ENSURE_AI(boss_four_horsemen::boss_four_horsemenAI, Sir->AI())->encounterActionAttack = true;
-
-                    ENSURE_AI(boss_four_horsemen::boss_four_horsemenAI, Thane->AI())->AttackStart(who);
-                    ENSURE_AI(boss_four_horsemen::boss_four_horsemenAI, Lady->AI())->AttackStart(who);
-                    ENSURE_AI(boss_four_horsemen::boss_four_horsemenAI, Baron->AI())->AttackStart(who);
-                    ENSURE_AI(boss_four_horsemen::boss_four_horsemenAI, Sir->AI())->AttackStart(who);
+                    TC_LOG_WARN("scripts", "FourHorsemenAI: Encounter starting but horseman with id %d is not present", uint32(boss));
+                    return ResetEncounter();
```

Comment:
please use "return;" when the return type is void otherwise it might look like the called function is returning some kind of value.